linalg/tensor.hpp

```diff
@@ -0,0 +1,2051 @@
+ *
+ * @note the different cases of 0D, 1D, 2D, 3D, and 4D are implemented separately
+ *       to work around a limitation in nvcc involving __host__ __device__ lambdas with `auto` parameters.
+ */
+MFEM_SUPPRESS_NVCC_HOSTDEVICE_WARNING
+template <int n1, int n2, int n3, typename lambda_type>
+MFEM_HOST_DEVICE auto make_tensor(lambda_type f) ->
+tensor<decltype(f(n1, n2, n3)), n1, n2, n3>
+{
+   using T = decltype(f(n1, n2, n3));
+   tensor<T, n1, n2, n3> A{};
+   for (int i = 0; i < n1; i++)
+   {
+      for (int j = 0; j < n2; j++)
+      {
+         for (int k = 0; k < n3; k++)
+         {
+            A(i, j, k) = f(i, j, k);
+         }
+      }
+   }
+   return A;
+}
+
+/**
+ * @brief Creates a tensor of requested dimension by subsequent calls to a functor
+ *
+ * @tparam n1 The first dimension of the tensor
+ * @tparam n2 The second dimension of the tensor
+ * @tparam n3 The third dimension of the tensor
+ * @tparam n4 The fourth dimension of the tensor
+ * @tparam lambda_type The type of the functor
+ * @param[in] f The functor to generate the tensor values from
+ * @pre @a f must accept @p n1 x @p n2 x @p n3 x @p n4 arguments of type @p int
+ *
+ * @note the different cases of 0D, 1D, 2D, 3D, and 4D are implemented separately
+ *       to work around a limitation in nvcc involving __host__ __device__ lambdas with `auto` parameters.
+ */
+MFEM_SUPPRESS_NVCC_HOSTDEVICE_WARNING
+template <int n1, int n2, int n3, int n4, typename lambda_type>
+MFEM_HOST_DEVICE auto make_tensor(lambda_type f) ->
+tensor<decltype(f(n1, n2, n3, n4)), n1, n2, n3, n4>
+{
+   using T = decltype(f(n1, n2, n3, n4));
+   tensor<T, n1, n2, n3, n4> A{};
+   for (int i = 0; i < n1; i++)
+   {
+      for (int j = 0; j < n2; j++)
+      {
+         for (int k = 0; k < n3; k++)
+         {
+            for (int l = 0; l < n4; l++)
+            {
+               A(i, j, k, l) = f(i, j, k, l);
+            }
+         }
+      }
+   }
+   return A;
+}
+
+/**
+ * @brief return the sum of two tensors
+ * @tparam S the underlying type of the lefthand argument
+ * @tparam T the underlying type of the righthand argument
+ * @tparam n integers describing the tensor shape
+ * @param[in] A The lefthand operand
+ * @param[in] B The righthand operand
+ */
+template <typename S, typename T, int... n>
+MFEM_HOST_DEVICE auto operator+(const tensor<S, n...>& A,
+                                const tensor<T, n...>& B) ->
+tensor<decltype(S {} + T{}), n...>
+{
+   tensor<decltype(S{} + T{}), n...> C{};
+   for (int i = 0; i < tensor<T, n...>::first_dim; i++)
+   {
+      C[i] = A[i] + B[i];
+   }
+   return C;
+}
+
+/**
+ * @brief return the unary negation of a tensor
+ * @tparam T the underlying type of the righthand argument
+ * @tparam n integers describing the tensor shape
+ * @param[in] A The tensor to negate
+ */
+template <typename T, int... n>
+MFEM_HOST_DEVICE tensor<T, n...> operator-(const tensor<T, n...>& A)
+{
+   tensor<T, n...> B{};
```

Comment:
Value init is redundant. Other similar unnecessary initializations below should be removed as well.